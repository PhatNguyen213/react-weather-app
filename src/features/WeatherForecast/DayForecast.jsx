import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import { CardActionArea } from '@mui/material';
import ThunderstormIcon from '@mui/icons-material/Thunderstorm';
import WbSunnyIcon from '@mui/icons-material/WbSunny';

const UnavailableForecast = () => (
  <Card sx={{ maxWidth: 345 }}>
    <CardActionArea>
      <CardContent>
        <Typography gutterBottom variant="h5" component="div">
          Unavailable
        </Typography>
      </CardContent>
    </CardActionArea>
  </Card>
);
const DayForecast = ({ forecast }) => {
  if (!forecast) return <UnavailableForecast />;
  const { date, day } = forecast;
  const { mintemp_c, maxtemp_c, daily_will_it_rain, condition } = day;
  const { text: summary } = condition;
  return (
    <Card sx={{ maxWidth: 345 }}>
      <CardActionArea>
        <CardMedia>
          {daily_will_it_rain ? <ThunderstormIcon /> : <WbSunnyIcon />}
        </CardMedia>
        <CardContent>
          <Typography gutterBottom variant="h5" component="div">
            {date}
          </Typography>
          <Typography gutterBottom variant="h5" component="div">
            {summary}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            Min Temperature: {mintemp_c}
          </Typography>
          <Typography variant="body2" color="text.secondary">
            Max Temperature: {maxtemp_c}
          </Typography>
        </CardContent>
      </CardActionArea>
    </Card>
  );
};

export default DayForecast;
