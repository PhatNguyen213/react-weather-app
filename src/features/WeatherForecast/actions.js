import { createAsyncThunk } from '@reduxjs/toolkit';
import { getWeatherForecastForLocation } from '../../api/services/weatherServices';

export const fetchWeatherForecastForLocation = createAsyncThunk(
  'weather/fetchWeatherForecastForLocation',
  async locationKey => {
    const { data } = await getWeatherForecastForLocation(locationKey);
    return data;
  }
);
