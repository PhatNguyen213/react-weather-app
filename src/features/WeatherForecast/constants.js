export const API_URLS = {
  GET_FORECASTS: '/forecast.json',
};
