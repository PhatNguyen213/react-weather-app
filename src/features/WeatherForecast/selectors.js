import { uniqueId } from 'lodash';
import { createSelector } from '@reduxjs/toolkit';

const forecastState = state => state.forecast;
const forecasts = createSelector(
  forecastState,
  state => state?.data?.forecast?.forecastday
);

const fillMissingDays = (arr, numOfMissingDays) => {
  const temp = [];
  for (let i = 1; i <= numOfMissingDays; i++) {
    temp.push({ date: uniqueId() });
  }
  return arr.concat(temp);
};

export const selectDaysForecasts = numberOfDays =>
  createSelector(forecasts, state => {
    if (!state) return [];
    if (state?.length === numberOfDays) return state;
    return fillMissingDays(state, numberOfDays - state?.length);
  });

export const selectIsFetching = createSelector(
  forecastState,
  state => state.loading
);

export const selectForecastLocationInfo = createSelector(
  forecastState,
  state => state?.data?.location
);
