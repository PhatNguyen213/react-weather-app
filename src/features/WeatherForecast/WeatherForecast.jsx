import DayForecast from './DayForecast';

const LocationSummary = (props = {}) => {
  const { location } = props;
  if (!location) return null;
  const { name, country, region, tz_id, localTime } = location;
  return (
    <div className="flex flex-col">
      <div>Name: {name}</div>
      <div>Country: {country}</div>
      <div>Region: {region}</div>
      <div>Timezone: {tz_id}</div>
      <div>Local time: {localTime}</div>
    </div>
  );
};

const Loading = () => <span>Loading...</span>;

const WeatherForecast = ({ forecasts, isFetching, location, city }) => {
  if (isFetching) return <Loading />;
  return (
    <>
      <LocationSummary location={location} />
      <div className="flex flex-row gap-4">
        {forecasts?.map(day => (
          <DayForecast key={day.date} forecast={day} />
        ))}
      </div>
    </>
  );
};

export default WeatherForecast;
