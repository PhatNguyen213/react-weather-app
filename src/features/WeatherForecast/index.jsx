import {
  selectDaysForecasts,
  selectForecastLocationInfo,
  selectIsFetching,
} from './selectors';
import { fetchWeatherForecastForLocation } from './actions';
import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import WeatherForecast from './WeatherForecast';

const select5DaysForecasts = selectDaysForecasts(5);

const WeatherForecastContainer = ({ city }) => {
  const dispatch = useDispatch();

  useEffect(() => {
    if (city) dispatch(fetchWeatherForecastForLocation(city));
  }, [dispatch, city]);

  const forecasts = useSelector(select5DaysForecasts);
  const isFetching = useSelector(selectIsFetching);
  const location = useSelector(selectForecastLocationInfo);
  return (
    <WeatherForecast
      forecasts={forecasts}
      isFetching={isFetching}
      location={location}
      city={city}
    />
  );
};

export default WeatherForecastContainer;
