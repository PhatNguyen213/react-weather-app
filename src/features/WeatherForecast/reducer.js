import { isAnyOf } from '@reduxjs/toolkit';
import { createReducer } from '@reduxjs/toolkit';
import { fetchWeatherForecastForLocation } from './actions';

const initWeatherState = {
  data: [],
  loading: false,
};

export default createReducer(initWeatherState, builder => {
  builder.addCase(
    fetchWeatherForecastForLocation.pending,
    (state, { payload }) => {
      state.loading = true;
    }
  );
  builder.addCase(
    fetchWeatherForecastForLocation.fulfilled,
    (state, { payload }) => {
      state.data = payload;
    }
  );
  builder.addMatcher(
    isAnyOf(
      fetchWeatherForecastForLocation.fulfilled,
      fetchWeatherForecastForLocation.rejected
    ),
    (state, { payload }) => {
      state.loading = false;
    }
  );
});
