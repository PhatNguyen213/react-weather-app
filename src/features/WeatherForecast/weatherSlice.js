import { API_URLS } from './constants';

const { GET_FORECASTS } = API_URLS;

export const weatherSlice = {
  endpoints: builder => ({
    getForecastsByLocationKey: builder.query({
      query: locationKey => ({
        url: `${GET_FORECASTS}`,
        params: {
          q: locationKey,
          days: 5,
        },
      }),
    }),
  }),
};
