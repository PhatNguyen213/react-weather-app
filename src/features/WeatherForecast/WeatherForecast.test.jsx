import { getAllByText, getByText } from '@testing-library/react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { selectDaysForecasts } from './selectors';
import WeatherForecast from './WeatherForecast';
import { isForecastAvailable } from './DayForecast';

const not = fn => args => !fn(args);

describe('WeatherForecast', () => {
  describe('renders correctly', () => {
    it('compare snapshots', () => {
      const { container } = render(
        <WeatherForecast
          forecasts={[
            {
              date: '2022-10-24',
              maxTemp: 17.4,
              minTemp: 12.1,
              avgTemp: 14.5,
              conditionSummary: 'Moderate rain',
              willItRain: 1,
            },
            {
              date: '2022-10-25',
              maxTemp: 12.4,
              minTemp: 12.1,
              avgTemp: 24.5,
              conditionSummary: 'Moderate rain',
              willItRain: 0,
            },
          ]}
          location={{
            name: 'London',
            region: 'City of London, Greater London',
            country: 'United Kingdom',
            tz_id: 'Europe/London',
            localTime: '2022-10-24 8:47',
          }}
          isFetching={false}
        />
      );
      expect(container).toMatchSnapshot();
    });
    it('shows loading status', () => {
      render(<WeatherForecast isFetching={true} />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });
    it('shows missing forecasts data as Unavailable', () => {
      const mockState = {
        forecast: {
          data: {
            forecasts: [
              {
                date: '2022-10-24',
                maxTemp: 17.4,
                minTemp: 12.1,
                avgTemp: 14.5,
                conditionSummary: 'Moderate rain',
                willItRain: 1,
              },
              {
                date: '2022-10-25',
                maxTemp: 12.4,
                minTemp: 12.1,
                avgTemp: 24.5,
                conditionSummary: 'Moderate rain',
                willItRain: 0,
              },
            ],
          },
        },
      };
      const select5DayForecasts = selectDaysForecasts(5);
      const forecasts = select5DayForecasts(mockState);
      render(<WeatherForecast forecasts={forecasts} isFetching={false} />);
      const unavailableForecasts = screen.getAllByText('Unavailable');
      expect(unavailableForecasts).toHaveLength(3);
    });
  });
  describe('selectors function correctly', () => {
    it('fill missing forecasts with unavailable data', () => {
      const mockState = {
        forecast: {
          data: {
            forecasts: [
              {
                date: '2022-10-24',
                maxTemp: 17.4,
                minTemp: 12.1,
                avgTemp: 14.5,
                conditionSummary: 'Moderate rain',
                willItRain: 1,
              },
              {
                date: '2022-10-25',
                maxTemp: 12.4,
                minTemp: 12.1,
                avgTemp: 24.5,
                conditionSummary: 'Moderate rain',
                willItRain: 0,
              },
            ],
          },
        },
      };
      const select5DayForecasts = selectDaysForecasts(5);
      const data = select5DayForecasts(mockState).filter(
        not(isForecastAvailable)
      );
      expect(data).toHaveLength(3);
    });
  });
});
