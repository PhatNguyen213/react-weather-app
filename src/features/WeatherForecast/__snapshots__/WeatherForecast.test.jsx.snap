// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherForecast renders correctly compare snapshots 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <div>
      Name: 
      London
    </div>
    <div>
      Country: 
      United Kingdom
    </div>
    <div>
      Region: 
      City of London, Greater London
    </div>
    <div>
      Timezone: 
      Europe/London
    </div>
    <div>
      Local time: 
      2022-10-24 8:47
    </div>
  </div>
  <div
    class="flex flex-row gap-4"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root grow css-1ri6ub7-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root flex justify-center css-46bh2p-MuiCardContent-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="ThunderstormIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17.92 7.02C17.45 4.18 14.97 2 12 2 9.82 2 7.83 3.18 6.78 5.06 4.09 5.41 2 7.74 2 10.5 2 13.53 4.47 16 7.5 16h10c2.48 0 4.5-2.02 4.5-4.5 0-2.34-1.79-4.27-4.08-4.48zM14.8 17l-2.9 3.32 2 1L11.55 24h2.65l2.9-3.32-2-1L17.45 17zm-6 0-2.9 3.32 2 1L5.55 24H8.2l2.9-3.32-2-1L11.45 17z"
            />
          </svg>
        </div>
        <div
          class="MuiCardContent-root flex flex-col items-center css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            2022-10-24
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            Moderate rain
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Min Temperature: 
            12.1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Max Temperature: 
            17.4
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root grow css-1ri6ub7-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root flex justify-center css-46bh2p-MuiCardContent-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
            data-testid="WbSunnyIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m6.76 4.84-1.8-1.79-1.41 1.41 1.79 1.79 1.42-1.41zM4 10.5H1v2h3v-2zm9-9.95h-2V3.5h2V.55zm7.45 3.91-1.41-1.41-1.79 1.79 1.41 1.41 1.79-1.79zm-3.21 13.7 1.79 1.8 1.41-1.41-1.8-1.79-1.4 1.4zM20 10.5v2h3v-2h-3zm-8-5c-3.31 0-6 2.69-6 6s2.69 6 6 6 6-2.69 6-6-2.69-6-6-6zm-1 16.95h2V19.5h-2v2.95zm-7.45-3.91 1.41 1.41 1.79-1.8-1.41-1.41-1.79 1.8z"
            />
          </svg>
        </div>
        <div
          class="MuiCardContent-root flex flex-col items-center css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            2022-10-25
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
          >
            Moderate rain
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Min Temperature: 
            12.1
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Max Temperature: 
            12.4
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
